### PR TITLE
Non-native field operations with ECC addition

### PR DESCRIPTION
**Pull request by mrain**
## Description

closes: #264 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Comment by mrain** [comment]
Brief explanation. Reference document: https://hackmd.io/@arielg/B13JoihA8

Let `E` be a larger prime field, and `F` be a smaller native prime field.
To emulate the field multiplication `a * b = c` in E, we will choose `T` such that `2^T * F::MODULUS > E::MODULUS ^ 2`,
then check `a * b - k * E::MODULUS = c (mod 2^T * F::MODULUS)` with witness `w`.
Using CRT, we can check the above equation mod both `F::MODULUS` and `2^T`. First one is easy, b/c `F` is the native field. We focus on `mod 2^T` case later.

First, note that we assume that `E::MODULUS > F::MODULUS`, so `2^T > E::MODULUS > F::MODULUS`. Storing elements in `E` modulo `2^T` is lossless. And for representing elements `mod 2^T` in `F`, we divide them into smaller limbs mod `2^B` where `2^{2B} * aux < F::MODULUS`, e.g. `a = a0 + a1 * 2^B + a2 * 2^{2B} + ...`. Then the multiplication mod `2^T` can be emulated over `F` without overflow.

**Review by chancharles92 — COMMENTED** [review]
Impressive job! @mrain 
Indeed not an easy PR.

File: relation/src/gadgets/ecc/non_native.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        let t2 = x2 * y1;
+        let d = d * t1 * t2;
+
+        let x3 = (t1 + t2) / (E::one() + d);
+        let y3 = (x1 * x2 + y1 * y2) / (E::one() - d);
+        let c = self.create_emulated_point_variable(Point(x3, y3))?;
+        self.emulated_ecc_add_gate(a, b, &c, d)?;
+        Ok(c)
+    }
+}
```
Comment (chancharles92) [inline relation/src/gadgets/ecc/non_native.rs:112]: Also add a test in this file.

File: relation/src/gadgets/emulated.rs [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+
+use crate::{errors::CircuitError, Circuit, PlonkCircuit, Variable};
+use ark_ff::PrimeField;
+use ark_std::{string::ToString, vec, vec::Vec, Zero};
+use core::marker::PhantomData;
+use itertools::izip;
+use num_bigint::BigUint;
+
+/// Parameters needed for emulating field operations over [`F`].
+pub trait EmulationConfig<F: PrimeField>: PrimeField {
```
[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs:21]: Just for simplicity. I use quite a lot `Into<BigUint>` in the implementation and it's not provided by `Field`. Also the application, `EmulatedPointVariable<E>` required `E` to be a `PrimeField` to instantiate a `Point<E>`.

File: relation/src/gadgets/emulated.rs [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+            self.enforce_in_range(next_carry_out, 1)?;
+
+            let wires = [var_k, *b, carry_out, next_carry_out, *c];
+            let coeffs = [a, F::one(), F::one(), -F::from(b_pow.clone())];
+            self.lc_gate(&wires, &coeffs)?;
+            carry_out = next_carry_out;
+
+            self.enforce_in_range(*b, E::B)?;
+        }
+        Ok(())
```
[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs:477]: Yes, `2^T` is already larger than `E::MODULUS + 1` so no need to check it `mod F::MODULUS`.

File: relation/src/gadgets/emulated.rs [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+        ))
+    }
+    /// Add a public emulated variable
+    pub fn create_public_emulated_variable<E: EmulationConfig<F>>(
+        &mut self,
+        val: E,
+    ) -> Result<EmulatedVariable<E>, CircuitError> {
+        let var = self.create_public_emulated_variable_unchecked(val)?;
+        for &v in &var.0 {
+            self.enforce_in_range(v, E::B)?;
```
[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs:145]: Public inputs are provided by users, I guess we should, right?

[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs:145]: Now I have another question, when should we do the range checks.
Doing it inside `add` and `mul` gates seems quite redundant. Should we always range check the variable when creating them?

[the remark this answers is not in the document]
Comment (chancharles92) [reply relation/src/gadgets/emulated.rs:136]: We should do range check when the variable to be range-checked is a secret witness that can't be accessed by the verifier. Agree that maybe we can add an API like `create_variable_with_range(..)`.

File: relation/src/gadgets/ecc/non_native.rs (hunk 1 above)
Comment (mrain) [reply relation/src/gadgets/ecc/non_native.rs:113]: Done in 9179f0c.

File: relation/src/gadgets/emulated.rs [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+            .map(|val| self.create_variable(val))
+            .collect::<Result<Vec<_>, CircuitError>>()?;
+
+        // Checking whether a + b = add_no_mod_limbs
+        let mut carry_out = self.zero();
+        for (a, b, c) in izip!(&a.0, &b.0, &add_no_mod_limbs) {
+            let next_carry_out =
+                F::from(<F as Into<BigUint>>::into(self.witness(*a)? + self.witness(*b)?) / &b_pow);
+            let next_carry_out = self.create_variable(next_carry_out)?;
+            self.enforce_in_range(next_carry_out, 1)?;
```
[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs]: Changed to `enforce_bool` in 69383e5

File: relation/src/gadgets/emulated.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs]: It's okay. The range check will be performed when `create_emulated_variables` now. 8c46301

File: relation/src/gadgets/emulated.rs [hunk 6]
```diff
@@ -0,0 +1,715 @@
...
+        ))
+    }
+    /// Add a constant emulated variable
+    pub fn create_constant_emulated_variable<E: EmulationConfig<F>>(
+        &mut self,
+        val: E,
+    ) -> Result<EmulatedVariable<E>, CircuitError> {
+        let var = self.create_constant_emulated_variable_unchecked(val)?;
+        for &v in &var.0 {
+            self.enforce_in_range(v, E::B)?;
```
[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs]: Removed in 69383e5

File: relation/src/gadgets/emulated.rs [hunk 7]
```diff
@@ -0,0 +1,715 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Emulate arithmetic operations on a large prime field.
+//! To emulate arithmetic operations on F_q when the native field is F_p where p
+//! < q, we represent the elements in F_q using CRT modulus [p, 2^T] where p *
+//! 2^T > q^2 + q. The second componenet, with modulus 2^T, will be divided into
```
[the remark this answers is not in the document]
Comment (mrain) [reply relation/src/gadgets/emulated.rs]: Added in 69383e5

File: relation/src/gadgets/ecc/non_native.rs (hunk 1 above)
Comment (mrain) [reply relation/src/gadgets/ecc/non_native.rs:113]: Done in [a47bdfb](https://github.com/EspressoSystems/jellyfish/pull/291/commits/a47bdfbb861be8507a5ccfc113e83f35112aa7ee)

File: relation/src/gadgets/emulated.rs [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+where
+    E: EmulationConfig<F>,
+    F: PrimeField,
+{
+    if vals.len() != E::NUM_LIMBS {
+        return Err(CircuitError::FieldAlgebraError(
+            "Malformed structure for emulated field element conversion.".to_string(),
+        ));
+    }
+    let b_pow = BigUint::from(2u32).pow(E::B as u32);
```
Comment (ggutoski) [inline relation/src/gadgets/emulated.rs:65]: same here with bit shifts

**Review by ggutoski — COMMENTED** [review]
I have nothing useful to add. Just a couple of trivial perf notes below.

**Review by chancharles92 — COMMENTED** [review]
LGTM, remember to sync with main branch

**Review by chancharles92 — COMMENTED** [review]
LGTM, remember to sync with main branch

**Review by chancharles92 — APPROVED** [review]

**Pull request merged**